### DISCUSSION TITLE
GCHead-less tuple.

### DISCUSSION THREAD

**Discussion by methane** [faster-cpython/ideas#304] [Ideas]
bytes, str, dict, and tuple are the four major memory eaters. I proposed three ideas to reduce bytes, str, dict memory usage:

* bytes -- remove ob_shash in Python 3.13
* str -- remove wstr cache in Python 3.12
* dict -- remove hash cache when all keys are unicode.

Now I am considering how to reduce memory usage of tuples. As you know, many tuples are created during loading modules (co_consts, co_names, etc...), and most of them are untracked because all contents are immutables. So GC header wastes 16bytes on each tuples.

I looked `_PyObject_IS_GC()` and found `tp_is_gc`. I never know it!

```c
// Fast inlined version of PyObject_IS_GC()
static inline int
_PyObject_IS_GC(PyObject *obj)
{
    return (PyType_IS_GC(Py_TYPE(obj))
            && (Py_TYPE(obj)->tp_is_gc == NULL
                || Py_TYPE(obj)->tp_is_gc(obj)));
}
```

I have three ideas how to implement tp_is_gc of tuple.

1. Use refcnt -- when object is immortal, treat it as no-GC tuple.
   * But this need "load module by immortal mode" option in addition to immortal PEP.
2. Use one bit of ob_size.
   * This will break `Py_SIZE(tuple)`.
3. Use separate heap and use its address to determine the tuple has GC header or not.
   * If we chose this option, all objects having GC header would be allocated on different heap.
   * This is tightly coupled with allocator. `PYTHONMALLOC=malloc valgrind ./python ...` would be difficult.

**Comment by JelleZijlstra** [comment]
What about putting the difference in `ob_type`? There could be a separate `PyNoNGCTuple_Type` that is exactly like a tuple except that it doesn't do GC.

I think that's how str objects work internally (https://docs.python.org/3.10/c-api/unicode.html#c.PyASCIIObject).

**Reply by methane** [reply]
> What about putting the difference in ob_type? There could be a separate PyNoNGCTuple_Type that is exactly like a tuple except that it doesn't do GC.

It is possible. Tuples like co_consts can be tuple-subclass, instead of exact tuple.
On the other hand, we can not use it for constants like `(1,2)`. It must be just tuple.

> I think that's how str objects work internally (https://docs.python.org/3.10/c-api/unicode.html#c.PyASCIIObject).

No. All Unicode structs has PyUnicode_Type.
Unicode uses PyASCIIObject.state to distinguish which structure is used.

**Comment by markshannon** [comment]
I would rather we just remove `tp_is_gc`.
We should be using declarative approaches rather than procedural ones, wherever possible.

**Reply by methane** [reply]
Then, how do you think GC module special casing tuples? Is it better than tp_is_gc?

GC module has some special logics for tuples already.

**Comment by sweeneyde** [comment]
3. Use separate heap and use its address to determine the tuple has GC header or not.
>    * If we chose this option, all objects having GC header would be allocated on different heap.
>    * This is tightly coupled with allocator. `PYTHONMALLOC=malloc valgrind ./python ...` would be difficult.

Is this portable enough? It's my understanding that comparing pointers from different buffers is implementation-defined and can give unexpected results on segmented address spaces: `ptr1<ptr2` might compare offsets but not the segment. I don't know how much those systems are supported.

[mimalloc](https://microsoft.github.io/mimalloc/group__analysis.html)'s mi_heap_check_owned() function is documented as

> Note: expensive function, linear in the pages in the heap.

**Reply by ericsnowcurrently** [reply]
> Is this portable enough? It's my understanding that comparing pointers from different buffers is implementation-defined and can give unexpected results on segmented address spaces: `ptr1<ptr2` might compare offsets but not the segment. I don't know how much those systems are supported.

Yeah, I've run into that part of the C spec too.  gt/lt on pointers is only defined for two pointers from the same struct/array.  So you AFAIK you can't efficiently tell if an arbitrary pointer is in a range.  Of course, there aren't that many platforms/compilers that rely on non-linear address spaces like that.  If we could easily identify them during preprocessing then we could avoid the linear search approach most of the time.

**Comment by methane** [comment]
I made a draft PR.
https://github.com/python/cpython/pull/31627

`PYTHONMALLOC=malloc valgrind --tool=massif ./python -c 'import asyncio'`

RAM used by tuple 706,096 bytes -> 393,712 (no GC header) + 211,792 (regular) bytes

<details>
<summary>main branch</summary>

```
89.54% (7,449,345B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->08.49% (706,096B) 0x411356: gc_alloc (gcmodule.c:2285)
| ->08.49% (706,096B) 0x411356: _PyObject_GC_NewVar (gcmodule.c:2320)
|   ->07.50% (624,216B) 0x2F1D31: tuple_alloc (tupleobject.c:96)
|   | ->07.50% (624,216B) 0x2F1D31: PyTuple_New (tupleobject.c:155)
|   |   ->05.52% (459,536B) 0x3D7877: r_object (marshal.c:1206)
|   |   | ->02.28% (189,640B) 0x3D82B3: r_object (marshal.c:1390)
|   |   | | ->02.24% (186,456B) 0x3D78A3: r_object (marshal.c:1212)
|   |   | | | ->02.24% (186,456B) 0x3D829D: r_object (marshal.c:1387)
|   |   | | |   ->01.50% (125,176B) 0x3D78A3: r_object (marshal.c:1212)
|   |   | | |   | ->01.50% (125,176B) 0x3D829D: r_object (marshal.c:1387)
|   |   | | |   |   ->01.45% (120,680B) 0x3D93F7: read_object (marshal.c:1524)
|   |   | | |   |   | ->01.45% (120,632B) 0x3D955C: marshal_loads_impl (marshal.c:1831)
|   |   | | |   |   | | ->01.45% (120,632B) 0x3D955C: marshal_loads (marshal.c.h:148)
|   |   | | |   |   | |   ->01.24% (102,896B) 0x22259C: _PyEval_EvalFrameDefault (ceval.c:4866)
|   |   | | |   |   | |   | ->01.24% (102,896B) 0x399C18: _PyEval_EvalFrame (pycore_ceval.h:53)
|   |   | | |   |   | |   |   ->01.24% (102,896B) 0x399C18: _PyEval_Vector (ceval.c:6238)
|   |   | | |   |   | |   |     ->01.24% (102,896B) 0x282D9B: _PyObject_VectorcallTstate (pycore_call.h:92)
|   |   | | |   |   | |   |       ->01.24% (102,896B) 0x282D9B: object_vacall (call.c:818)
|   |   | | |   |   | |   |         ->01.24% (102,896B) 0x285377: PyObject_CallMethodObjArgs (call.c:878)
|   |   | | |   |   | |   |           ->01.24% (102,896B) 0x3CC0A9: import_find_and_load (import.c:1729)
|   |   | | |   |   | |   |             ->01.24% (102,896B) 0x3CC0A9: PyImport_ImportModuleLevelObject (import.c:1828)
|   |   | | |   |   | |   |               ->01.01% (83,792B) 0x227612: import_name (ceval.c:7200)
|   |   | | |   |   | |   |               | ->01.01% (83,792B) 0x227612: _PyEval_EvalFrameDefault (ceval.c:3871)
|   |   | | |   |   | |   |               |   ->01.01% (83,792B) 0x399ABC: _PyEval_EvalFrame (pycore_ceval.h:53)
|   |   | | |   |   | |   |               |     ->01.01% (83,792B) 0x399ABC: _PyEval_Vector (ceval.c:6238)
|   |   | | |   |   | |   |               |       ->01.01% (83,792B) 0x399ABC: PyEval_EvalCode (ceval.c:1101)
|   |   | | |   |   | |   |               |         ->01.01% (83,792B) 0x395BE8: builtin_exec_impl (bltinmodule.c:1040)
|   |   | | |   |   | |   |               |           ->01.01% (83,792B) 0x395BE8: builtin_exec (bltinmodule.c.h:371)
|   |   | | |   |   | |   |               |             ->01.01% (83,792B) 0x2DBA42: cfunction_vectorcall_FASTCALL (methodobject.c:427)
|   |   | | |   |   | |   |               |               ->01.01% (83,792B) 0x28399D: PyObject_Call (call.c:348)
|   |   | | |   |   | |   |               |                 ->01.01% (83,792B) 0x221BEA: do_call_core (ceval.c:7128)
|   |   | | |   |   | |   |               |                   ->01.01% (83,792B) 0x221BEA: _PyEval_EvalFrameDefault (ceval.c:5195)
|   |   | | |   |   | |   |               |                     ->01.01% (83,792B) 0x399C18: _PyEval_EvalFrame (pycore_ceval.h:53)
|   |   | | |   |   | |   |               |                       ->01.01% (83,792B) 0x399C18: _PyEval_Vector (ceval.c:6238)
|   |   | | |   |   | |   |               |                         ->01.01% (83,792B) 0x282D9B: _PyObject_VectorcallTstate (pycore_call.h:92)
|   |   | | |   |   | |   |               |                           ->01.01% (83,792B) 0x282D9B: object_vacall (call.c:818)
|   |   | | |   |   | |   |               |                             ->01.01% (83,792B) 0x285377: PyObject_CallMethodObjArgs (call.c:878)
|   |   | | |   |   | |   |               |                               ->01.01% (83,792B) 0x3CC0A9: import_find_and_load (import.c:1729)
|   |   | | |   |   | |   |               |                                 ->01.01% (83,792B) 0x3CC0A9: PyImport_ImportModuleLevelObject (import.c:1828)
|   |   | | |   |   | |   |               |                                   ->01.01% (83,792B) in 2 places, all below massif's threshold (1.00%)
|   |   | | |   |   | |   |               |
|   |   | | |   |   | |   |               ->00.23% (19,104B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | | |   |   | |   |
|   |   | | |   |   | |   ->00.21% (17,736B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | | |   |   | |
|   |   | | |   |   | ->00.00% (48B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | | |   |   |
|   |   | | |   |   ->00.05% (4,496B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | | |   |
|   |   | | |   ->00.74% (61,280B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | | |
|   |   | | ->00.04% (3,184B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | |
|   |   | ->01.55% (129,000B) 0x3D829D: r_object (marshal.c:1387)
|   |   | | ->01.52% (126,064B) 0x3D78A3: r_object (marshal.c:1212)
|   |   | | | ->01.52% (126,064B) 0x3D829D: r_object (marshal.c:1387)
|   |   | | |   ->01.52% (126,064B) in 2 places, all below massif's threshold (1.00%)
|   |   | | |
|   |   | | ->00.04% (2,936B) in 1+ places, all below ms_print's threshold (01.00%)
|   |   | |
|   |   | ->01.28% (106,832B) 0x3D82CC: r_object (marshal.c:1393)
|   |   | | ->01.28% (106,832B) 0x3D78A3: r_object (marshal.c:1212)
|   |   | |   ->01.28% (106,832B) 0x3D829D: r_object (marshal.c:1387)
|   |   | |     ->01.28% (106,832B) in 2 places, all below massif's threshold (1.00%)
```
</details>

<details>
<summary>gc-less tuple</summary>

```
--------------------------------------------------------------------------------
  n        time(i)         total(B)   useful-heap(B) extra-heap(B)    stacks(B)
--------------------------------------------------------------------------------
 58    192,411,986        8,229,368        7,359,578       869,790            0
89.43% (7,359,578B) (heap allocation functions) malloc/new/new[], --alloc-fns, etc.
->04.78% (393,712B) 0x2DF5F6: _PyObject_NewVar (object.c:195)
| ->04.78% (393,688B) 0x2F2710: _PyTuple_NewNoGC (tupleobject.c:180)
| | ->04.78% (393,688B) 0x3D8757: r_object (marshal.c:1210)
| |   ->02.00% (164,992B) 0x3D91EB: r_object (marshal.c:1403)
| |   | ->02.00% (164,368B) 0x3D8786: r_object (marshal.c:1216)
| |   | | ->02.00% (164,368B) 0x3D91D5: r_object (marshal.c:1400)
| |   | |   ->01.43% (118,088B) 0x3D8786: r_object (marshal.c:1216)
| |   | |   | ->01.43% (118,088B) 0x3D91D5: r_object (marshal.c:1400)
| |   | |   |   ->01.39% (114,648B) 0x3DA367: read_object (marshal.c:1537)
| |   | |   |   | ->01.39% (114,648B) 0x3DA4CC: marshal_loads_impl (marshal.c:1844)
| |   | |   |   | | ->01.39% (114,648B) 0x3DA4CC: marshal_loads (marshal.c.h:148)
| |   | |   |   | |   ->01.23% (101,472B) 0x22351B: _PyEval_EvalFrameDefault (ceval.c:4866)
| |   | |   |   | |   | ->01.23% (101,472B) 0x39A9F0: _PyEval_EvalFrame (pycore_ceval.h:53)
| |   | |   |   | |   |   ->01.23% (101,472B) 0x39A9F0: _PyEval_Vector (ceval.c:6238)
| |   | |   |   | |   |     ->01.23% (101,472B) 0x282EEB: _PyObject_VectorcallTstate (pycore_call.h:92)
| |   | |   |   | |   |       ->01.23% (101,472B) 0x282EEB: object_vacall (call.c:818)
| |   | |   |   | |   |         ->01.23% (101,472B) 0x285577: PyObject_CallMethodObjArgs (call.c:878)
| |   | |   |   | |   |           ->01.23% (101,472B) 0x3CCF89: import_find_and_load (import.c:1729)
| |   | |   |   | |   |             ->01.23% (101,472B) 0x3CCF89: PyImport_ImportModuleLevelObject (import.c:1828)
| |   | |   |   | |   |               ->01.23% (101,472B) in 2 places, all below massif's threshold (1.00%)
| |   | |   |   | |   |
| |   | |   |   | |   ->00.16% (13,176B) in 1+ places, all below ms_print's threshold (01.00%)
| |   | |   |   | |
| |   | |   |   | ->00.00% (0B) in 1+ places, all below ms_print's threshold (01.00%)
| |   | |   |   |
| |   | |   |   ->00.04% (3,440B) in 1+ places, all below ms_print's threshold (01.00%)
| |   | |   |
| |   | |   ->00.56% (46,280B) in 1+ places, all below ms_print's threshold (01.00%)
| |   | |
| |   | ->00.01% (624B) in 1+ places, all below ms_print's threshold (01.00%)
| |   |
| |   ->01.27% (104,104B) 0x3D91D5: r_object (marshal.c:1400)
| |   | ->01.26% (103,448B) 0x3D8786: r_object (marshal.c:1216)
| |   | | ->01.26% (103,448B) 0x3D91D5: r_object (marshal.c:1400)
| |   | |   ->01.26% (103,448B) in 2 places, all below massif's threshold (1.00%)
| |   | |
| |   | ->00.01% (656B) in 1+ places, all below ms_print's threshold (01.00%)
| |   |
| |   ->01.23% (101,024B) 0x3D9204: r_object (marshal.c:1406)
| |   | ->01.23% (101,024B) 0x3D8786: r_object (marshal.c:1216)
| |   |   ->01.23% (101,024B) 0x3D91D5: r_object (marshal.c:1400)
| |   |     ->01.23% (101,024B) in 2 places, all below massif's threshold (1.00%)
|
->02.57% (211,792B) 0x412356: gc_alloc (gcmodule.c:2285)
| ->02.57% (211,792B) 0x412356: _PyObject_GC_NewVar (gcmodule.c:2320)
|   ->01.79% (146,896B) 0x2F2631: tuple_alloc (tupleobject.c:94)
|   | ->01.79% (146,896B) 0x2F2631: PyTuple_New (tupleobject.c:152)
|   |   ->01.79% (146,896B) in 24 places, all below massif's threshold (1.00%)
|   |
|   ->00.79% (64,896B) in 1+ places, all below ms_print's threshold (01.00%)
```

**Reply by methane** [reply]
pyperformance

<details>
<summary><code>$ ./python -m pyperf compare_to ../main.json tuple.json -G --table --table-format=md</code></summary>

| Benchmark               | main    | tuple                 |
|-------------------------|:-------:|:---------------------:|
| pidigits                | 234 ms  | 219 ms: 1.07x faster  |
| scimark_sparse_mat_mult | 7.02 ms | 6.61 ms: 1.06x faster |
| pickle_dict             | 37.8 us | 35.7 us: 1.06x faster |
| nbody                   | 114 ms  | 110 ms: 1.04x faster  |
| regex_effbot            | 3.95 ms | 3.82 ms: 1.04x faster |
| unpack_sequence         | 57.4 ns | 55.5 ns: 1.03x faster |
| regex_dna               | 239 ms  | 232 ms: 1.03x faster  |
| pathlib                 | 26.5 ms | 25.8 ms: 1.03x faster |
| sqlalchemy_imperative   | 27.7 ms | 27.0 ms: 1.03x faster |
| pyflate                 | 574 ms  | 560 ms: 1.02x faster  |
| spectral_norm           | 148 ms  | 145 ms: 1.02x faster  |
| regex_v8                | 28.9 ms | 28.2 ms: 1.02x faster |
| telco                   | 8.92 ms | 8.73 ms: 1.02x faster |
| meteor_contest          | 128 ms  | 126 ms: 1.02x faster  |
| scimark_lu              | 152 ms  | 149 ms: 1.02x faster  |
| unpickle_pure_python    | 342 us  | 336 us: 1.02x faster  |
| go                      | 187 ms  | 183 ms: 1.02x faster  |
| unpickle_list           | 5.89 us | 5.80 us: 1.02x faster |
| hexiom                  | 8.60 ms | 8.49 ms: 1.01x faster |
| scimark_sor             | 159 ms  | 157 ms: 1.01x faster  |
| scimark_fft             | 467 ms  | 462 ms: 1.01x faster  |
| mako                    | 13.8 ms | 13.7 ms: 1.01x faster |
| crypto_pyaes            | 109 ms  | 108 ms: 1.01x faster  |
| sympy_expand            | 676 ms  | 671 ms: 1.01x faster  |
| sympy_str               | 408 ms  | 405 ms: 1.01x faster  |
| sympy_integrate         | 27.2 ms | 27.1 ms: 1.01x faster |
| chaos                   | 96.8 ms | 96.3 ms: 1.00x faster |
| logging_silent          | 151 ns  | 150 ns: 1.00x faster  |
| sympy_sum               | 226 ms  | 225 ms: 1.00x faster  |
| python_startup          | 10.2 ms | 10.2 ms: 1.00x slower |
| fannkuch                | 495 ms  | 498 ms: 1.01x slower  |
| python_startup_no_site  | 7.27 ms | 7.31 ms: 1.01x slower |
| nqueens                 | 114 ms  | 115 ms: 1.01x slower  |
| json_loads              | 34.5 us | 34.8 us: 1.01x slower |
| logging_format          | 9.48 us | 9.56 us: 1.01x slower |
| xml_etree_parse         | 180 ms  | 182 ms: 1.01x slower  |
| 2to3                    | 327 ms  | 330 ms: 1.01x slower  |
| sqlite_synth            | 3.11 us | 3.15 us: 1.01x slower |
| django_template         | 50.6 ms | 51.2 ms: 1.01x slower |
| unpickle                | 18.8 us | 19.1 us: 1.02x slower |
| richards                | 66.6 ms | 68.0 ms: 1.02x slower |
| pickle_list             | 5.31 us | 5.43 us: 1.02x slower |
| pickle                  | 13.5 us | 13.8 us: 1.03x slower |
| chameleon               | 8.96 ms | 9.21 ms: 1.03x slower |
| Geometric mean          | (ref)   | 1.01x faster          |

Benchmark hidden because not significant (15): html5lib, sqlalchemy_declarative, scimark_monte_carlo, tornado_http, raytrace, dulwich_log, pickle_pure_python, json_dumps, deltablue, xml_etree_process, regex_compile, xml_etree_generate, xml_etree_iterparse, logging_simple, float

</details>
